core/src/main/java/io/grpc/internal/AbstractServerStream.java

```diff
@@ -189,10 +189,15 @@ protected TransportState(int maxMessageSize, StatsTraceContext statsTraceCtx) {
      * thread.
      */
     public final void setListener(ServerStreamListener listener) {
+      Preconditions.checkState(this.listener == null, "setListener should be called only once");
       this.listener = Preconditions.checkNotNull(listener, "listener");
+    }
 
-      // Now that the stream has actually been initialized, call the listener's onReady callback if
-      // appropriate.
+    /**
+     * Once the stream has actually been initialized, call the listener's onReady callback if
+     * appropriate.
+     */
+    public final void listenerReady() {
```

Comment:
Go ahead and make onStreamAllocated non-`final` and then redefine it as public here:
```java
public final void onStreamAllocated() {
  super.onStreamAllocated();
}
```

We do something similar with `AbstractStream2.isReady()` and some other cases.